### PR TITLE
aotf: niceify menu and pre-filter mutations

### LINKED ISSUE #621: aotf: shorter mutation lists
The design document shows menus with nice short lists of commands:

Screenshot 2021-03-19 at 15 36 59

However, we have quite a few mutations that could be called in different contexts so the list is quite a lot longer and we still have all of those offline mutations and potentially user-defined ones too #558.

We should try to present a shorter list of more relevant commands with the option to expand to see more. (This was actually a part of the early design sketches and run by a few user groups.)

Ideas to reduce the number of commands:

1) A relevance score based on the number of arguments that relate to Cylc Objects (flows, cycles, tasks, etc) that cannot be satisfied by the context.

    * E.G. If you click on a workflow the trigger mutation would be considered to have a low relevance because it takes tasks as arguments and we are clicking on a workflow not a task.
    * I think this logic would go in `@/utils/aotf:filterMutations`.

2) Pre-defined short-lists.

    * E.G. hard-coded lists of commands to display by default for different Cylc Object types.

Perhaps try (1) and fall-back to (2)?

From the design perspective this would me a "more" button at the end of the list or something like that?

### PR DESCRIPTION
**Pull request by oliver-sanders**
closes #621 
addresses #622

Before:
Screenshot 2021-03-31 at 15 22 22

After:
Screenshot 2021-03-31 at 15 22 27

Lessons learned:

* The `dense` prop doesn't really help with space efficiency and makes the fonts go nasty.
* Easier to get layout/style right without using the `v-list` component, however, hard to get the hover effects standard without it.

Screengrab:

Kapture 2021-03-31 at 17 03 21

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [ ] Appropriate change log entry included.
- [x] No documentation update required.

### CODE REVIEW

**Comment by hjoliver** [comment]
Very nice :+1:

**Comment by kinow** [comment]
> Very nice +1

+1 on that! Looks great!

File: CHANGES.md [hunk 1]
```diff
@@ -16,6 +16,10 @@ None or N/A.
 
 ### Enhancements
 
+[#636](https://github.com/cylc/cylc-ui/pull/636) -
+Improve the appearence of mutation menus and display an expandable short list
```
Comment (kinow) [inline CHANGES.md:20]: s/appearence/appearance ? (Not sure if this spelling exists too in us/gb english)

**Review by kinow — COMMENTED** [review]
@oliver-sanders I'm trying to build it locally, but I think I am either doing something silly (probably due to the long leave :coffee: ) or the PR is out of sync/date with your local code?

Here's what I have when I build it and run behind Cylc UI Server.

image

I've already nuked my venv's and my `node_module` folders. Rebuilt everything. Then removed my `dist` folder in Cylc UI. Cleared browser cache. But it keeps displaying the longer list.

I had a look at the code, and it looks like you are using `c-menu` in `aotf.scss`, but I couldn't find it in any of our files? There's `c-mutation-menu` in `Menu.vue`, but changing it to `c-menu` still didn't show me the new menu layout? :confused:

File: CHANGES.md (hunk 1 above)
Comment (hjoliver) [reply CHANGES.md:20]: no, it is an error

**Comment by oliver-sanders** [comment]
> @oliver-sanders I'm trying to build it locally, but I think I am either doing something silly (probably due to the long leave ☕ ) or the PR is out of sync/date with your local code?

Have you configured the UIS to use your local build rather than the bundled one? Here's my config:

```python
# ~/.cylc/hub/config.py
from pathlib import Path
c.UIServer.ui_build_dir = Path('~/cylc-ui/dist').expanduser()
```

**Comment by kinow** [comment]
> > @oliver-sanders I'm trying to build it locally, but I think I am either doing something silly (probably due to the long leave coffee ) or the PR is out of sync/date with your local code?
> 
> Have you configured the UIS to use your local build rather than the bundled one? Here's my config:
> 
> ```python
> # ~/.cylc/hub/config.py
> from pathlib import Path
> c.UIServer.ui_build_dir = Path('~/cylc-ui/dist').expanduser()
> ```
> 
> You can see where the UI is being served from in the `cylc hub` log lines that are written when the UIS is started.

:man_facepalming: I did create a config file, but I didn't check if it was working. Most likely that's why I was seeing no difference, sorry. Will try it again tomorrow morning.

File: src/components/cylc/cylcObject/Menu.vue [hunk 2]
```diff
@@ -158,11 +202,12 @@ export default {
...
     },
 
-    showMutationsMenu ({ id, types, tokens, mutations, event }) {
+    showMutationsMenu ({ id, type, types, tokens, mutations, event }) {
       this.id = id
+      this.type = type
       this.types = types
       this.tokens = tokens
-      this.mutations = mutations
+      this.mutations = mutations.sort()
```
[the remark this answers is not in the document]
Comment (kinow) [reply src/components/cylc/cylcObject/Menu.vue:210]: Ah! I think I see what you wanted to do here. Here's the list of workflow mutations I'm seeing:

Screenshot from 2021-04-08 17-47-24

But here's what I think this PR was aiming at?

image

[the remark this answers is not in the document]
Comment (kinow) [reply src/components/cylc/cylcObject/Menu.vue:210]:
```suggestion
      this.mutations = mutations.sort((a, b) => a[0].name.localeCompare(b[0].name))
```

[the remark this answers is not in the document]
Comment (kinow) [reply src/components/cylc/cylcObject/Menu.vue:210]: The `localeCompare` is probably an overkill, we won't have mutations named like `10`, `1`, etc, I think :man_shrugging:

File: CHANGES.md (hunk 1 above)
Comment (hjoliver) [inline CHANGES.md:20]:
```suggestion
Improve the appearance of mutation menus and display an expandable short list
```

Comment (hjoliver) [reply CHANGES.md:20]: (make it easy to action @kinow 's suggestion :grin: )

Comment (hjoliver) [reply CHANGES.md:20]: (screw it, I'll apply that change).

**Review by hjoliver — APPROVED** [review]
Tested, LGTM. @kinow 's questions still to be addressed.

**Comment by hjoliver** [comment]
(Note this now needs master merged in to test, since the host -> platform schema change).

**Comment by oliver-sanders** [comment]
Re-ran the tests (GH tests against a merge commit with master).

I want to sort out the sorting (see Bruno's comments above), the intention was that mutations would appear in the order defined in `utils/aotf` (i.e. play, pause, reload, stop).

**Comment by oliver-sanders** [comment]
Sorted, the mutations should now appear in the hardcoded order (i.e. `play`, `pause`, `reload`, `stop`).

File: src/components/cylc/cylcObject/Menu.vue [hunk 3]
```diff
@@ -135,17 +144,56 @@ export default {
...
+      const shortList = this.$workflowService.primaryMutations[this.type]
+      if (!this.expanded && shortList) {
+        return this.mutations.filter(
+          (x) => {
+            return shortList.includes(x[0].name)
+          }
+        ).sort(
+          (x, y) => {
+            return shortList.indexOf(x[0].name) - shortList.indexOf(y[0].name)
+          }
```
Comment (kinow) [inline src/components/cylc/cylcObject/Menu.vue:171]: So default order is alphabetical, but `primaryMutations` list can specify another order for the mutations to be displayed :+1:

**Review by kinow — APPROVED** [review]

**Comment by kinow** [comment]
Two approvals, LGTM

**Pull request merged**